### PR TITLE
Allow custom executors without indirection

### PR DESCRIPTION
**Pull request by bugadani**
This pull request replaces the `Pender` enum with an `extern "Rust"` callback that must be implemented by architectures/platforms to support embassy-executor

### CODE REVIEW

**Comment by bugadani** [comment]
I am still considering a few things:
 - Whether or not I should revert the ThreadModeExecutor changes. Reimplementing isn't a big deal, but on the other hand, neither is implementing a trait instead of the whole executor facade.
 - I'm not particularly fond of the API. The interrupt-mode executor receives its context in `start`, the thread-mode in its constructor.
 - The thread-mode executor looks like it could be a single function that creates the executor and runs it. There seems to be no reason to provide a static object that can't be used for anything else.
 - As an alternative to the above point, option B is to add the same `spawner` function to the thread-mode executor like what `InterruptExecutor` has, so that the static global can be useful, and the API can become more symmetrical :)

**Comment by Dirbaio** [comment]
(for context, original discussion in Matrix [here](https://matrix.to/#/!LdaNPfUfvefOLewEIM:matrix.org/$JWsn5EU4gR1zq84BbkYFP8rwBY-koBP_Nj1pOU3yHQ8?via=matrix.org&via=tchncs.de&via=mozilla.org))

Hmm, so what you're doing here is keeping the `Pender` enum with the "thread, interrupt, callback" cases, but making customizable what the "thread" and "interrupt" cases do via `extern "Rust"`. I think it's a bit "worst of both worlds": we have *both* the enum which was a bit of a hack, *and also* the `extern` trick, both of which come with their associated complexity. Let me explain:

Before #1321 the `raw` module was completely agnostic to the "thread vs interrupt mode executors". It just took a function pointer for the pender. The non-raw Executor and InterruptExecutor would supply function pointers that did the right thing (SEV, or pend an irq).

In #1321 I added the `Pender` enum as sort of a hack to avoid indirect calls when using the builtin thread-mode / interrupt-mode executors. This meant `raw` is no longer agnostic to the executor kind. However, I still tried to keep `raw`'s public API looking "as if" it was agnostic. I wanted to keep the non-agnosticness as an implementation detail, as just an optimization. This is why there's no way to construct `Pender::Thread` or `Pender::Interrupt` from ouside the crate, for example. 

Ideally `raw` should be agnostic to "thread-mode, interrupt-mode" because it's supposed to be arch-agnostic. Some arch might have "frob-mode, glorb-mode, zlosh-mode" instead. 

I think if we do add a way to customize the pender via `extern "Rust"`, the enum is redundant, because you can already do everything without it. In the Matrix discussion, I mentioned that as a design goal we should have "allow mixing different Penders in the same binary (like one thread executor + N interrupt executors)". However, that's not entirely true: if you can't mix penders, you can still set a single pender that does an `if`, so different executor instances do different things on pend. So, my proposal:

- Remove the `Pender` enum.
- Make `raw` look like before #1321, except the function pointer is now an `extern "Rust"` fn. That is:
   - You can set a *single* pender callback via `extern "Rust"`. All raw executor instances will call that.
   - You must pass a `pender_ctx: *mut ()` to `raw::Executor::new()`. It is passed to the pender callback.
- all `arch-xxx` set the pender callback extern fn to do the arch-specific thing.
- `arch-cortex-m` sets a *single* pender callback that does `if pender_ctx as usize == 0xFFFF_FFFF { sev() } else { NVIC.pend(pender_ctx as irqn) }`.

If no `arch-xxx` feature is enabled, the user has to set their own pender callback (or use a lib that does it) or they get a linker error.

Advantages:
- `raw` is truly arch-agnostic again
- Everything works with no indirect calls
- If the user still wants indirect calls, they can still do it by setting a pender callback that calls a function pointer.
- External crates are on the same "footing" to implement executor flavors as the embassy-executor crate. All work with no indirect calls.

Disadvantages:

- It's no longer possible to "mix" a builtin executor (like from `arch-cortex-m`) and a custom one. If you go custom, you have to go full custom.

----

as an aside, to answer why the executor API looks like it does:

> The thread-mode executor looks like it could be a single function that creates the executor and runs it. There seems to be no reason to provide a static object that can't be used for anything else.

the executor needs to live forever. If the executor is stack-allocated in `run()`, it's possible to exit from it with panic + catch_unwind, which would deallocate it. Could be fixed by making the executor itself do `catch_uwnind` in the right places, but it's not trivial.

> As an alternative to the above point, option B is to add the same spawner function to the thread-mode executor like what InterruptExecutor has, so that the static global can be useful, and the API can become more symmetrical :)

There's some necessary asymmetry: `InterruptExecutor::spawner()` returns a `SendSpawner`, which is more restrictive (it requires task args to be Send). It's nice to be able to obtain a (non-Send) `Spawner` for the thread mode executor, so you can move non-Send things from the non-async main to the tasks via args. For that, the compiler has to be able to enforce that you're obtaining the spawner and calling `run()` on the same thread. For this, `Executor` has to be `!Send`, so you can't place it in a static unlike `InterruptExecutor`...

**Comment by bugadani** [comment]
I have written a lengthy response, in which I managed to convince myself that you are right :)

`pender-callback` implementors will have to deal with only receiving a single pointers worth of context, but that is probably an okay compromise for the reduced complexity in embassy.

---

> It's nice to be able to obtain a (non-Send) Spawner for the thread mode executor, so you can move non-Send things from the non-async main to the tasks via args

I'm not sure I understand this sentence correctly, but I did phrase my thoughts badly. What I meant, was a `Executor::spawner()` that returns a `SendSpawner` so that interrupt executors can spawn tasks on the thread executor (or one core could spawn tasks on another). Those tasks should already be Send, I think, as they would be passing a thread boundary, but I may be wrong. Generalizing to a non-send `Spawner` sounds way less useful because one can just get one via `Spawner::for_current_executor()`.

> the executor needs to live forever

I think all hell breaks loose already if a program unwinds from the run function (I don't think it's possible to reset tasks, and that means calling `run` again will just probably panic again on any `must_spawn` call, and so nothing else can be done with the executor after unwinding from it), but this is a fair point I didn't consider.

**Comment by Dirbaio** [comment]
> I think all hell breaks loose already if a program unwinds from the run function

Yes, the executor becomes "useless" but at least it doesn't create unsoundness.

> I'd rather if the context was a usize.

from the "strict provenance" discussions in Rust lately, this has been discussed a bit. Nothing is formalized yet, but the consensus is smuggling a `usize` through a pointer is fine (as long as you don't deref it of course) but smuggling a pointer through a `usize` is not. It strips provenance info, and in some architectures such as ARM CHERI, pointers are physically "address + more stuff", so you can't restore it from a `usize` which is just the address.

**Comment by bugadani** [comment]
> "address + more stuff"

eh, right, I forgot about this one... anyway, easy fix if we can have `PenderContext` opaque.

**Comment by Dirbaio** [comment]
I think it'd be simpler to outright remove `Pender` and `PenderContext`. There's no advantage in making it opaque, and it'd get rid of all the transmutes.

Also I'd remove `interrupt.rs` for now. Whether we want an arch-agnostic InterruptExecutor or not is a separate question, we can discuss later. I'd focus on refactoring the pender only in this PR.

**Comment by bugadani** [comment]
`PenderContext` wasn't entirely pointless, but the idea needed the agnostic thread-mode executor. The idea was to disallow arbitrary values passed as context - the platform implementation should have defined the supported conversions.

I'll resolve the compile errors soon and then this should be ready, I think?

File: embassy-executor/src/arch/cortex_m.rs [hunk 1]
```diff
@@ -1,3 +1,48 @@
...
+#[cfg(any(feature = "executor-thread", feature = "executor-interrupt"))]
+fn __pender(context: *mut ()) {
+    unsafe {
+        // Safety: `context` is either `usize::MAX` created by `Executor::run`, or a valid interrupt
+        // request number given to `InterruptExecutor::start`.
+
+        let context = context as usize;
+
+        #[cfg(feature = "executor-thread")]
+        if context == THREAD_PENDER {
```
[the remark this answers is not in the document]
Comment (bugadani) [reply embassy-executor/src/arch/cortex_m.rs:13]: Yeah, I originally had 3 different functions for this reason...

**Comment by bugadani** [comment]
Converting to draft because of the unupdated docs.

**Review by Dirbaio — APPROVED** [review]
*Very* nice. I'm so glad to hae the `Pender` enum removed, the design is looking much better now than before. Hopefully it covers all use cases this time! Thank you for working on this.

**Pull request merged**